CONTRIBUTING.md

```diff
@@ -474,3 +483,127 @@ that should be reported, so there's no need to make these messages readable to t
+
+[Nextflow only works with ADC.](https://www.nextflow.io/docs/latest/google.html#credentials) User is asked to run `gcloud auth application-default login`, which writes  `.config/gcloud/application_default_credentials.json`.
+
+In Docker mode, [we mount `.config/gcloud`](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/main/java/bio/terra/cli/app/DockerCommandRunner.java#L80).
+
+</td>
+</tr>
+<tr>
+<td>GCP notebook</td>
+<td>
+
+**Pet ADC** (if you use `--use-app-default-credentials`)
+
+Notebooks can run terra CLI with  [`--use-app-default-credentials`](https://github.com/DataBiosphere/terra-cli/pull/197/files).
+terra will pick up creds from ADC instead of OAuth (which involves interacting with a browser).
+</td>
+<td valign="top">
+
+**Pet ADC**
+
+ADC is automatically configured for GCE VMs.
+</td>
+</tr>
+<tr>
+<td>Unit test</td>
+<td valign="top">
+
+**Test user Terra OAuth**
+
+At the beginning of each test, before running terra CLI, [test user is logged in](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/test/java/harness/baseclasses/SingleWorkspaceUnit.java#L30).
+
+`TestUser.login()` [writes `.terra/StoredCredential`](https://github.com/DataBiosphere/terra-cli/blob/8adf7cdaaa1f74f9407c10cccc8f7c0c4623eb6b/src/test/java/harness/TestUser.java#L76), using domain-wide delegation to avoid browser flow.
+</td>
+<td>
+
+**Test user**
+
+Auth is complicated because we are not using Pet SA key (which is a security risk).
+
+*Access token*
+
+TestCommand.runCommand(): Set `TEST_USER_ACCESS_TOKEN` system property
+
+DockerCommand: If `TEST_USER_ACCESS_TOKEN` system property is set, set `CLOUDSDK_AUTH_ACCESS_TOKEN` to access token in container.
+(Tests only run in docker mode.)
+
+*bq*
+
+Here's what makes `cloudsdk/component_build/wrapper_scripts/bq.py` happy:
+- Set `CLOUDSDK_AUTH_ACCESS_TOKEN` to access token
+- Populate `.config/gcloud/legacy_credentials/default/adc.json`. This file has refresh token.
+
+In general for gcloud, you only need `--access_token_file`/`CLOUDSDK_AUTH_ACCESS_TOKEN`, not `adc.json`.
+However, [implementation of `CLOUDSDK_AUTH_ACCESS_TOKEN`](https://cloud.google.com/sdk/docs/release-notes#cloud_sdk_2) did not include `bq.py`, so we also need `adc.json`.
+
+*gsutil*
+
+There are 3 versions of `gsutil`:
+
+1. Standalone `gsutil`. This is the oldest, and predates `gcloud`.
+2. `gsutil` as part of gcloud
+3. `gcloud alpha storage`. This is the newest. [Faster than 2.](https://stackoverflow.com/collectives/google-cloud/articles/68475140/faster-cloud-storage-transfers-using-the-gcloud-command-line)
+
+This repo uses 2. To get auth to work for 2, we write the `~/.config/gcloud/legacy_credentials/default/.boto` that  `google-cloud-sdk/bin/bootstrapping/gsutil.py` expects. This file has refresh token.
+
+For 3, auth is simpler. You just set `CLOUDSDK_AUTH_ACCESS_TOKEN`  (which we are already setting for other gcloud pass-through commands).
```

Comment:
Do we use `gcloud alpha storage` in this repo or is this noting a future improvement?